doc/developer/design/20210505_kafka_keys_in_sql.md

```diff
@@ -0,0 +1,165 @@
+
+* An explicit key/value format:
+
+  ```sql
+  KEY FORMAT <specifier>
+  VALUE FORMAT <specifier>
+  ```
+
+* A bare format that specifies a Confluent Schema Registry (CSR) config
+
+  ```sql
+  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY ...
+  ```
+
+If `INCLUDE KEY` is specified, the source schema will be augmented to prepend
+all of the values of the key as columns in the dataflow. Format types that do
+not specify subfields (e.g. `FORMAT TEXT`) will use their standard column names
+(i.e. `text`).
+
+Alternatively, if the `INCLUDE KEY AS RECORD <name>` syntax is used all fields
+will be included as fields on a record with the specified name.
+
+If any key column names colide with value columns using the bare `INCLUDE KEY`
+syntax, an error will be raised before dataflow construction suggesting the use
+of the `AS RECORD` clause.
+
+### Implementation
+
+Mostly standard for new syntax additions. We will need to wait until
+`purify_format` to provide any of the specific error-detection because of the
+fact that Key schemas may come from the Confluent Schema Registry.
+
+Physically unpacking key fields will happen identically independent of if `AS
+RECORD` is specified, only the projection mapping and names available to SQL
+will differ.
+
+## Alternatives
+
+<!--
+// Similar to the Description section. List of alternative approaches considered, pros/cons or why they were not chosen
+-->
+
+### Alternative: Just docs
+
+We could just improve documentation around getting Kafka keys into the value
+section using Kafka itself.
+
+### Alternative: Record-valued function
+
+It might be possible to provide something like a `kafka_key()` function that
+introduces a demand for the key, pulling it into dataflows.
+
+### Alternative: key and value records
+
+Instead of automatically extracting Key columns at all, we could provide a
+syntax like:
+
+```sql
+CREATE SOURCE <source-name> FROM KAFKA BROKER '<>'
+<format>
+[AS KEY VALUE]
+```
+
+which which would change emit the Kafka message key as a `key` record, but would
+*also* change the message value to be available as a `value` record (no longer
+automatically expanding the fields into columns).
+
+Downsides to this approach are that a small bit of syntax dramatically changes
+usage patterns from the SQL layer. That has effects both on documentation which
+would now have to deal with a niche alternative usage. Changing to `AS KEY
+VALUE` would force a rewrite or intermediate view if a user realizes that they
+need data from the key after already using a topic for awhile.
+
+## Open questions
+
+<!--
+// Anything currently unanswered that needs specific focus. This section may be expanded during the doc meeting as
+// other unknowns are pointed out.
+// These questions may be technical, product, or anything in-between.
+-->
+
+We could make the `AS RECORD <name>` syntax required. The proposed design is
+based on the assumption that the common case is that key fields do not conflict
```

Comment:
My expectation is that column name conflicts will be quite frequent, as noted above.

Reply:
right, let's keep the conversation on that part.